File: q101-125/q121.py
# 2020.02.09
# Problem Statement:
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

# I still remember this is exactly the first question I got in my interview with Intel. I did it pretty bad. I was, and I still am such a loser :-(
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # initialize buy and sell day and profit
        buy_day, sell_day, profit = 0, 0, 0

        for i in range(0, len(prices)):
            # the larger value of price would contribute to more profit with the current buy day, but not necessarily an overall larger profit
            if prices[i] > prices[sell_day]:
                sell_day = i
                profit = max(profit, prices[i] - prices[buy_day])
            
            # the smaller value could be the new buy day, but the sell day should be changed accordingly
            elif prices[i] < prices[buy_day]:
                sell_day, buy_day = i, i
            
        return profit

File: Amazon/q121.java
// 2022.09.01
// Problem Statement:
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

// idea: each new day after day one could be the sell day (update answer) or the buy day (update curr buy day)
class Solution {
    public int maxProfit(int[] prices) {
        int answer = 0;
        int curr_buy_day = 0;
        for (int i=1; i<prices.length; i++) {
            // could be the sell day
            answer = Math.max(answer, prices[i]-prices[curr_buy_day]);
            
            // could be the buy day
            if (prices[i]<=prices[curr_buy_day]) {
                curr_buy_day = i;
            }
        }
        return answer;
    }
}